File: Tamrin TA/1.py
from collections import defaultdict 

class Graph(): 
	def __init__(self, V): 
		self.V = V 
		self.graph = defaultdict(list) 

	def addEdge(self, u, v): 
		self.graph[u].append(v) 

	def DFSUtil(self, u, color): 
		color[u] = "GRAY"
		for v in self.graph[u]: 

			if color[v] == "GRAY": 
				return True

			if color[v] == "WHITE" and self.DFSUtil(v, color) == True: 
				return True

		color[u] = "BLACK"
		return False

	def isCyclic(self): 
		color = ["WHITE"] * self.V 

		for i in range(self.V): 
			if color[i] == "WHITE": 
				if self.DFSUtil(i, color) == True: 
					return True
		return False

edge_count=int(input()) 
g=Graph(edge_count)

for i in range(edge_count):
    data=input()
    data=list(map(int, data.split()))
    g.addEdge(data[0],data[1])
    data.clear()

if(g.isCyclic() == True):
    print(1)
else:
    print(0)
							

File: Tamrin TA/1.cpp
#include <bits/stdc++.h> 
using namespace std; 

enum Color {WHITE, GRAY, BLACK}; 

class Graph 
{ 
	int V; // No. of vertices 
	list<int>* adj; // adjacency lists 

	bool DFSUtil(int v, int color[]); 
public: 
	Graph(int V); // Constructor 
	void addEdge(int v, int w); 

	bool isCyclic(); 
}; 

Graph::Graph(int V) 
{ 
	this->V = V; 
	adj = new list<int>[V]; 
} 

void Graph::addEdge(int v, int w) 
{ 
	adj[v].push_back(w);  
} 

bool Graph::DFSUtil(int u, int color[]) 
{ 

	color[u] = GRAY; 

	list<int>::iterator i; 
	for (i = adj[u].begin(); i != adj[u].end(); ++i) 
	{ 
		int v = *i; 


		if (color[v] == GRAY) 
		return true; 

		if (color[v] == WHITE && DFSUtil(v, color)) 
		return true; 
	} 


	color[u] = BLACK; 

	return false; 
} 
 
bool Graph::isCyclic() 
{ 
	int *color = new int[V]; 
	for (int i = 0; i < V; i++) 
		color[i] = WHITE; 

	for (int i = 0; i < V; i++) 
		if (color[i] == WHITE) 
		if (DFSUtil(i, color) == true) 
			return true; 

	return false; 
} 
 
int main() 
{ 
    int n,a,b;
    cin>>n;
    
	Graph g(n); 
	for(int i=0;i<n;i++)
	{
	    cin >> a >> b;
	    g.addEdge(a, b);
	}

	if (g.isCyclic()) 
		cout <<"1"; 
	else
		cout << "0"; 

	return 0; 
} 
